src/misc/programs.cpp

```diff
@@ -170,75 +189,52 @@ bool Program::SuppressWriteOut(const char *format)
 	return true;
 }
 
-// For "\n" to "\r\n" expansion (0xA to OxD 0xA) in WriteOut* functions
-static char last_written_character = '\n';
+// TODO: Refactor messages and related functions like WriteOut that use variadic
+// args to instead use format strings when dosbox-staging starts using C++20
 
-void Program::WriteOut(const char *format, ...)
+void Program::WriteOut(const char* format, ...)
 {
-	if (SuppressWriteOut(format))
+	if (SuppressWriteOut(format)) {
 		return;
+	}
 
-	char buf[2048];
-	va_list msg;
+	char buf[WRITEOUT_BUFSIZE];
+	std::va_list msg;
 
-	va_start(msg,format);
-	vsnprintf(buf,2047,format,msg);
+	va_start(msg, format);
+	int size = std::vsnprintf(buf, WRITEOUT_BUFSIZE, format, msg) + 1;
 	va_end(msg);
```

Comment:
Programmatically, we never want strings > 2000, so the too-long-so-we-truncate scenario should be more like an exception or a full blown failure.

I realize the code is safe and defensive (and that's good!).

But I think we should `assert(size < WRITEOUT_BUFSIZE);` straight-away, so debug builds stop in their tracks, and developers can fix the problem.

For release builds, we should `LOG_ERR(..)` too, just to have a loud and visual log of the problematic string.

Reply:
Oh, that's good to know: I didn't realize this is processing user-side data and arguments. In that case, yes, I fully agree with the approach.